crates/dbsp/src/operator/trace.rs

```diff
@@ -230,6 +276,145 @@ where
         trace.clone()
     }
 
+    /// Like `integrate_trace`, but additionally applies a retainment policy to
+    /// keys in the trace.
+    ///
+    /// ## Background
+    ///
+    /// Relations that store time series data typically have the property that
+    /// any new updates can only affect records with recent timestamps.
+    /// Depending on how the relation is used in queries this might mean
+    /// that, while records with older timestamps still exist in the
+    /// relation, they cannot affect any future incremental computation and
+    /// therefore don't need to be stored.
+    ///
+    /// ## Design
+    ///
+    /// We support two mechanism to specify and eventually discard such unused
+    /// records.
+    ///
+    /// The first mechanism, exposed via the
+    /// [`integrate_trace_with_bound`](`Self::integrate_trace_with_bound`)
+    /// method, is only applicable when keys and/or values in the collection
+    /// are ordered by time.  It allows _each_ consumer of the trace to specify
+    /// a lower bound on the keys and values it is interested in.  The
+    /// effective bound is the minimum of all bounds specified by individual
+    /// consumers.
+    ///
+    /// The second mechanism, implemented by this method and the
+    /// [`integrate_trace_retain_values`](`Self::integrate_trace_retain_values`)
+    /// method, is more general and allows the caller to specify an
+    /// arbitrary condition on keys and values in the trace respectively.
```

Comment:
it's really a function that is applied to keys or values and which is controlled by outside parameters.
but it must have the property that it's monotone, in that a key or value once rejected is always rejected

Reply:
Added a clause about this.